README.md
# react-native-google-fit

[Gitter Group](https://gitter.im/React-native-google-fit/community) - ask questions, answer questions!

[![npm version](https://badge.fury.io/js/react-native-google-fit.svg)](https://badge.fury.io/js/react-native-google-fit) ![Downloads](https://img.shields.io/npm/dm/react-native-google-fit.svg)

A React Native bridge module for interacting with Google Fit

# Quick Links
- [Installation](/docs/INSTALLATION.md)
- [Changelog](/docs/CHANGELOG.md)
- [Example app](https://github.com/StasDoskalenko/react-native-google-fit-example)
- [F.A.Q.](/docs/FAQ.md)


### USAGE

1. `import GoogleFit, { Scopes } from 'react-native-google-fit'`

2. Authorize:

    To check whethere GoogleFit is already authorized, simply use a function
    ```
        GoogleFit.checkIsAuthorized()
    ```
    Then you can simply refer to `GoogleFit.isAuthorized` boolean.
    
    ```javascript
    // The list of available scopes inside of src/scopes.js file
    const options = {
      scopes: [
        Scopes.FITNESS_ACTIVITY_READ_WRITE,
        Scopes.FITNESS_BODY_READ_WRITE,
      ],
    }
    GoogleFit.authorize(options)
      .then(authResult => {
        if (authResult.success) {
          dispatch("AUTH_SUCCESS");
        } else {
          dispatch("AUTH_DENIED", authResult.message);
        }
      })
      .catch(() => {
        dispatch("AUTH_ERROR");
      })
    
    // ...
    // Call when authorized
    GoogleFit.startRecording((callback) => {
      // Process data from Google Fit Recording API (no google fit app needed)
    });
    ```
    
    Alternatively you can use event listeners (deprecated)
    ```javascript
     GoogleFit.onAuthorize(() => {
       dispatch('AUTH SUCCESS')
     })
         
     GoogleFit.onAuthorizeFailure(() => {
       dispatch('AUTH ERROR')
     })
    ```

3. Retrieve Steps For Period

    ```javascript
    const options = {
      startDate: "2017-01-01T00:00:17.971Z", // required ISO8601Timestamp
      endDate: new Date().toISOString() // required ISO8601Timestamp
    };
    
    GoogleFit.getDailyStepCountSamples(options)
     .then((res) => {
         console.log('Daily steps >>> ', res)
     })
     .catch((err) => {console.warn(err)})
    
    // shortcut functions, 
    // return weekly or daily steps of given date
    // all params are optional, using new Date() without given date, 
    // adjustment is 0 by default, determine the first day of week, 0 == Sunday, 1==Monday, etc.
    GoogleFit.getDailySteps(date).then.catch()
    GoogleFit.getWeeklySteps(date, adjustment).then().catch()
     
    ```

    **Response:**
    
    ```javascript
    [
      { source: "com.google.android.gms:estimated_steps", steps: [
        {
          "date":"2019-06-29","value":2328
        },
        {
          "date":"2019-06-30","value":8010
          }
        ]
      },
      { source: "com.google.android.gms:merge_step_deltas", steps: [
        {
          "date":"2019-06-29","value":2328
        },
        {
          "date":"2019-06-30","value":8010
          }
        ]
      },
      { source: "com.xiaomi.hm.health", steps: [] }
    ];
    ```

4. Retrieve Weights

    ```javascript
    const opt = {
      unit: "pound", // required; default 'kg'
      startDate: "2017-01-01T00:00:17.971Z", // required
      endDate: new Date().toISOString(), // required
      ascending: false // optional; default false
    };
    
    GoogleFit.getWeightSamples(opt, (err, res) => {
      console.log(res);
    });
    ```

    **Response:**
    
    ```javascript
    [
      {
        "addedBy": "app_package_name",
        "value":72,
        "endDate":"2019-06-29T15:02:23.413Z",
        "startDate":"2019-06-29T15:02:23.413Z",
        "day":"Sat"
      },
      {
        "addedBy": "app_package_name",
        "value":72.4000015258789,
        "endDate":"2019-07-26T08:06:42.903Z",
        "startDate":"2019-07-26T08:06:42.903Z",
        "day":"Fri"
      }
    ]
    ```

5. Retrieve Heights

    ```javascript
    const opt = {
      startDate: "2017-01-01T00:00:17.971Z", // required
      endDate: new Date().toISOString(), // required
    };
    
    GoogleFit.getHeightSamples(opt, (err, res) => {
      console.log(res);
    });
    ```

    **Response:**

    ```javascript
    [
      {
        "addedBy": "app_package_name",
        "value":1.7699999809265137,
        "endDate":"2019-06-29T15:02:23.409Z",
        "startDate":"2019-06-29T15:02:23.409Z",
        "day":"Sat"
      }
    ]
    ```

6. Save Weights

    ```javascript
    const opt = {
      value: 200,
      date: new Date().toISOString(),
      unit: "pound"
    };
    
    GoogleFit.saveWeight(opt, (err, res) => {
      if (err) throw "Cant save data to the Google Fit";
    });
    ```
    
7. Blood pressure and Heart rate methods (since version 0.8)
    ```javascript
    const options = {
      startDate: "2017-01-01T00:00:17.971Z", // required
      endDate: new Date().toISOString(), // required
    }
    const callback = ((error, response) => {
      console.log(error, response)
    });

    GoogleFit.getHeartRateSamples(options, callback)
    GoogleFit.getBloodPressureSamples(options, callback)
    ```

    **Response:**

    ```javascript
    // heart rate
    [
      {
        "value":80,
        "endDate":"2019-07-26T10:19:21.348Z",
        "startDate":"2019-07-26T10:19:21.348Z",
        "day":"Fri"
      }
    ]

    // blood pressure
    [
      {
        "value":120,
        "value2":80,
        "endDate":"2019-07-26T08:39:28.493Z",
        "startDate":"1970-01-01T00:00:00.000Z",
        "day":"Thu"
      }
    ]
    ```

8. Get all activities
    ```javascript
      let options = {
        startDate: new Date(2018, 9, 17).valueOf(), // simply outputs the number of milliseconds since the Unix Epoch
        endDate: new Date(2018, 9, 18).valueOf()
      };
      GoogleFit.getActivitySamples(options, (err, res) => {
        console.log(err, res)
      });
    ```

      **Response:**
    
    ```javascript
     [ { 
      sourceName: 'Android',
      device: 'Android',
      sourceId: 'com.google.android.gms',
      calories: 764.189208984375,
      quantity: 6,
      end: 1539774300992,
      tracked: true,
      activityName: 'still',
      start: 1539727200000 },
    { sourceName: 'Android',
      device: 'Android',
      sourceId: 'com.google.android.gms',
      calories: 10.351096153259277,
      quantity: 138,
      end: 1539774486088,
      tracked: true,
      distance: 88.09545135498047,
      activityName: 'walking',
    }]
    ```
    Where:
    ```
    sourceName = device - 'Android' or 'Android Wear' string
    sourceId - return a value of dataSource.getAppPackageName(). For more info see: https://developers.google.com/fit/android/data-attribution
    start/end - timestamps of activity in format of milliseconds since the Unix Epoch
    tracked - bool flag, is this activity was entered by user or tracked by device. Detected by checking milliseconds of start/end timestamps. Since when user log activity in googleFit they can't set milliseconds
    distance(opt) - A distance in meters.
    activityName - string, equivalent one of these https://developers.google.com/fit/rest/v1/reference/activity-types 
    calories(opt) - double value of burned Calories in kcal.
    quantity(opt) - equivalent of steps number
    ```
    Note that optional parametrs are not presented in all activities - only where google fit return some results for this field.
    Like no distance for still activity. 

9. Retrieve Calories For Period
    ```javascript
      const opt = {
        startDate: "2017-01-01T00:00:17.971Z", // required
        endDate: new Date().toISOString(), // required
        basalCalculation: true, // optional, to calculate or not basalAVG over the week
      };

      GoogleFit.getDailyCalorieSamples(opt, (err, res) => {
        console.log(res);
      });
    ```

    **Response:**
    
    ```javascript
    [
      {
        "calorie":1721.948974609375,
        "endDate":"2019-06-27T15:13:27.000Z",
        "startDate":"2019-06-27T15:02:23.409Z",
        "day":"Thu"
      },
      {
        "calorie":1598.25,
        "endDate":"2019-06-28T15:13:27.000Z",
        "startDate":"2019-06-27T15:13:27.000Z",
        "day":"Thu"
      }
    ]
    ```

10. Retrieve Distance For Period:
    ```javascript
      const opt = {
        startDate: "2017-01-01T00:00:17.971Z", // required
        endDate: new Date().toISOString(), // required
      };

      GoogleFit.getDailyDistanceSamples(opt, (err, res) => {
        console.log(res);
      });
    ```

    **Response:**
    
    ```javascript
    [
      {
        "distance":2254.958251953125,
        "endDate":"2019-06-30T15:45:32.987Z",
        "startDate":"2019-06-29T16:57:01.047Z",
        "day":"Sat"
      },
      {
        "distance":3020.439453125,
        "endDate":"2019-07-01T13:08:31.332Z",
        "startDate":"2019-06-30T16:58:44.818Z",
        "day":"Sun"
      }
    ]
    ```

11. Retrieve Daily Nutrition Data for Period:
    ```javascript
      const opt = {
        startDate: "2017-01-01T00:00:17.971Z", // required
        endDate: new Date().toISOString(), // required
      };

      GoogleFit.getDailyNutritionSamples(opt, (err, res) => {
        console.log(res);
      });
    ```

    **Response:**
    
    ```javascript
    [
      {
        "nutrients":{"sugar":14,"sodium":1,"calories":105,"potassium":422},
        "date":"2019-07-02"
      },
      {
        "nutrients":{"sugar":36,"iron":0,"fat.saturated":3.6000001430511475,"sodium":0.13500000536441803,"fat.total":6,"calories":225,"fat.polyunsaturated":0,"carbs.total":36,"potassium":0.21000000834465027,"cholesterol":0.029999999329447746,"protein":9.299999237060547},
        "date":"2019-07-25"
      }
    ]
    ```

15. Other methods:

    ```javascript
    observeSteps(callback); // On Step Changed Event

    unsubscribeListeners(); // Put into componentWillUnmount() method to prevent leaks

    isAvailable(callback); // Checks is GoogleFit available for current account / installed on device
    
    isEnabled(callback); // Checks is permissions granted
    
    deleteWeight(options, callback); // method to delete weights by options (same as in delete hydration)
 
    openFit(); //method to open google fit app
    
    saveHeight(options, callback);
 
    deleteHeight(options, callback); // method to delete heights by options (same as in delete hydration)
 
    disconnect(); // Closes the connection to Google Play services.
    ```

### PLANS / TODO

* code refactoring
* optimization

Copyright (c) 2017-present, Stanislav Doskalenko
doskalenko.s@gmail.com

Based on Asim Malik android source code, copyright (c) 2015, thanks mate!


docs/CHANGELOG.md
### Changelog:

```
0.10.0  
        ~ Fix weight and height deletion
        ~ Update typescript definitions

0.9.17  ~ observeSteps fix (kudos to @nojas01)
        ~ Fix disconnect method (kudos to @AylanBoscarino)
        ~ better React Native 0.60 support (kudos to @spacekadet)

0.9.15  + getUserInputSteps function (kudos @HelloCore)
        + Retrieve Daily Nutrition Data (kudos @jguix) 

0.9.13  + Improve weights granularity and always use FIELD_AVERAGE (@chrisgibbs44)

0.9.12  ~ Update typescript definitions for Scoped Authorization

0.9.11  ~ getDailyCalorieSamples, now includes basalCalculation: boolean flag

0.9.10-beta
        + Scope authorizations (thanks, @gaykov)
        
0.9.1   ~ getDailyStepCountSamples - now returns promise if no callback is provided
0.9.0   ~ authorize() - is now a promise
        ~ non-blocking step retrieve
        + getHeartRateSamples (thanks @damnnkst)
        + getBloodPressureSamples (thanks @damnnkst)
0.7.1   - Fix for disconnect() (@dmitriys-lits thanks for the PR)
0.7     - Retrieve Heights, open fit activity, unified body method (@EJohnF thanks for the PR!)

0.6     - RN 0.56+ support (@skb1129 thanks for the PR)
        - nutrition scenario (@13thdeus thanks for the PR)
        
0.5     - New auth process (@priezz thanks for PR)
        - Fix unsubscribe listeners
        - Readme refactoring
        
0.4.0-beta
        - Recording API implemetation (@reboss thanks for PR)
        - Just use startRecording(callback) function which listens
        to STEPS and DISTANCE (for now) activities from Google Fitness
        API (no need to install Google fit app)

0.3.5   - Fix Error: Fragments should be static
        - Updated readme

0.3.4   - Burned Calories History (getDailyCalorieSamples)

0.3.2
        - React Native 0.46 Support

0.3.1-beta
        - better cancel/deny support

0.3.0-beta (@firodj thanks for this PR!)
        - steps adapter to avoid errors;
        - authorize: allow cancel;
        - authorize: using callback instead event;
        - strict dataSource;
        - xiaomi support;

0.2.0   - getDailyDistanceSamples();
        - isAvailable();
        - isEnabled();
        - deleteWeight();
0.1.1-beta
        - getDailyStepCountSamples method compatible with Apple Healthkit module
        - started to implement JSDoc documentation

0.1.0
        - getting activity within module itself
        - fixed package name dependency
        - provided more detailed documentation

0.0.9   - Weights Save Support
        - Refactor methods to be compatible with react-native-apple-healthkit module
        - Remove 'moment.js' dependency

0.0.8   - Weights Samples support

0.0.1   - 0.0.7 Initial builds
```


index.android.js
'use strict'
import { DeviceEventEmitter, NativeModules, PermissionsAndroid } from 'react-native';

import PossibleScopes from './src/scopes';
import {
  buildDailySteps,
  isNil,
  KgToLbs,
  lbsAndOzToK,
  prepareDailyResponse,
  prepareResponse,
  prepareDeleteOptions,
  getWeekBoundary,
} from './src/utils';

const googleFit = NativeModules.RNGoogleFit

class RNGoogleFit {
  eventListeners = []
  isAuthorized = false

  authorize = async (options = {}) => {
    const successResponse = { success: true }
    try {
      await this.checkIsAuthorized()
      if (this.isAuthorized) {
        return successResponse
      }
      const authResult = await new Promise((resolve, reject) => {
        this.onAuthorize(() => {
          this.isAuthorized = true
          resolve(successResponse)
        })
        this.onAuthorizeFailure(error => {
          this.isAuthorized = false
          reject({ success: false, message: error.message })
        })

        const defaultScopes = [
          Scopes.FITNESS_ACTIVITY_READ,
          Scopes.FITNESS_BODY_READ_WRITE,
          Scopes.FITNESS_LOCATION_READ,
        ]

        googleFit.authorize({
          scopes: (options && options.scopes) || defaultScopes,
        })
      })
      return authResult
    } catch (error) {
      return { success: false, message: error.message }
    }
  }

  checkIsAuthorized = async () => {
    const { isAuthorized } = await googleFit.isAuthorized()
    this.isAuthorized = isAuthorized
  }

  disconnect = () => {
    this.isAuthorized = false
    googleFit.disconnect()
    this.removeListeners()
  }

  removeListeners = () => {
    this.eventListeners.forEach(eventListener => eventListener.remove())
    this.eventListeners = []
  }


  // recommend to refactor both permission to allow other permission options besides PERMISSONS.ACCESS_FINE_LOCATION
  // check permissions
  checkPermissionAndroid = async () => {
    const response = await PermissionsAndroid.check(PermissionsAndroid.PERMISSIONS.ACCESS_FINE_LOCATION);
    return response === true;
  }

  // request permissions
  requestPermissionAndroid = async (dataTypes) => {

    const check = await this.checkPermissionAndroid();

    if (dataTypes.includes('distance') && !check) {
      try {
        const granted = await PermissionsAndroid.request(
          PermissionsAndroid.PERMISSIONS.ACCESS_FINE_LOCATION,
          {
            title: "Access Location Permisson",
            message:
              "Enable location access for Google Fit Api. " +
              "Cancel may cause inaccuray result",
            buttonNegative: "Cancel",
            buttonPositive: "OK"
          }
        );

        // this need to be changed in the future if we want to use RecordingAPI for more sensitive permissions
        if( granted === PermissionsAndroid.RESULTS.GRANTED ) {
          // we don't do anything here since the permissons are granted
        } else {
          // remove distance from array to avoid crash,
          return dataTypes.filter(data => data !== 'distance');
        }
      } catch (err) {
        console.warn(err);
      };
    }
    return dataTypes;
  }

  /**
   * Start recording fitness data
   *
   * You could specify data by array dataTypes. Possible values - step, distance, activity which corresponds
   * DataTypes.TYPE_STEP_CUMULATIVE, DataType.TYPE_DISTANCE_DELTA and DataType.TYPE_ACTIVITIES_SAMPLES
   *
   * Default value for dataTypes is steps and distance data
   *
   * This function relies on sending events to signal the RecordingAPI status
   * Simply create an event listener for the {DATA_TYPE}_RECORDING (ex. STEP_RECORDING)
   * and check for {recording: true} as the event data
   */
  startRecording = (callback, dataTypes = ['step']) => {

    this.requestPermissionAndroid(dataTypes).then((dataTypes) => {
      googleFit.startFitnessRecording(dataTypes)

      const eventListeners = dataTypes.map(dataTypeName => {
        const eventName = `${dataTypeName.toUpperCase()}_RECORDING`

        return DeviceEventEmitter.addListener(eventName, event => callback(event))
      })

      this.eventListeners.push(...eventListeners)
    })
  }


  /**
   * A shortcut to get the total steps of a given day by using getDailyStepCountSamples
   * @param {Date} date optional param, new Date() will be used if date is not provided
   */
  getDailySteps(date = new Date()) {
    const options = {
      startDate: new Date(date.setHours(0,0,0,0)).toISOString(), // required ISO8601Timestamp
      endDate: new Date(date.setHours(23,59,59,999)).toISOString(), // required ISO8601Timestamp
    };
    return this.getDailyStepCountSamples(options);
  }

  /**
   * A shortcut to get the weekly steps of a given day by using getDailyStepCountSamples
   * @param {Date} date optional param, new Date() will be used if date is not provided
   * @param {number} adjustment, use to adjust the default start day of week, 0 = Sunday, 1 = Monday, etc.
   */
  getWeeklySteps(date=new Date(), adjustment=0) {
    const [startDate, endDate] = getWeekBoundary(date, adjustment);
    const options = {
      startDate: startDate,
      endDate: endDate,
    }
    return this.getDailyStepCountSamples(options);
  }

  _retrieveDailyStepCountSamples = (startDate, endDate, callback) => {
    googleFit.getDailyStepCountSamples(
      startDate,
      endDate,
      msg => callback(msg, false),
      res => {
        if (res.length > 0) {
          callback(
            false,
            res.map(function(dev) {
              const obj = {}
              obj.source =
                dev.source.appPackage +
                (dev.source.stream ? ':' + dev.source.stream : '')
              obj.steps = buildDailySteps(dev.steps)
              return obj
            }, this)
          )
        } else {
          callback('There is no any steps data for this period', false)
        }
      }
    )
  }

  /**
   * Get the total steps per day over a specified date range.
   * @param {Object} options getDailyStepCountSamples accepts an options object containing required startDate: ISO8601Timestamp and endDate: ISO8601Timestamp.
   * @param {Function} callback The function will be called with an array of elements.
   */

  getDailyStepCountSamples = (options, callback) => {
    const startDate = !isNil(options.startDate)
      ? Date.parse(options.startDate)
      : new Date().setHours(0, 0, 0, 0)
    const endDate = !isNil(options.endDate)
      ? Date.parse(options.endDate)
      : new Date().valueOf()
    if (!callback || typeof callback !== 'function') {
      return new Promise((resolve, reject) => {
        this._retrieveDailyStepCountSamples(
          startDate,
          endDate,
          (error, result) => {
            if (!error) {
              resolve(result)
            } else {
              reject(error)
            }
          }
        )
      })
    }
    this._retrieveDailyStepCountSamples(startDate, endDate, callback)
  }

  /**
   * Get the total steps per day over a specified date range.
   * @param {Object} options getUserInputSteps accepts an options object containing required startDate: ISO8601Timestamp and endDate: ISO8601Timestamp.
   * @param {Function} callback The function will be called with an array of elements.
   */

  getUserInputSteps = (options, callback) => {
    const startDate = !isNil(options.startDate) ? Date.parse(options.startDate) : (new Date()).setHours(0, 0, 0, 0)
    const endDate = !isNil(options.endDate) ? Date.parse(options.endDate) : (new Date()).valueOf()
    googleFit.getUserInputSteps(startDate, endDate,
      (msg) => callback(msg, false),
      (res) => {
        callback(null, res);
      }
    )
  }

  /**
   * Get the total distance per day over a specified date range.
   * @param {Object} options getDailyDistanceSamples accepts an options object containing required startDate: ISO8601Timestamp and endDate: ISO8601Timestamp.
   * @param {function} callback The function will be called with an array of elements.
   */

  getDailyDistanceSamples(options, callback) {
    const startDate = !isNil(options.startDate)
      ? Date.parse(options.startDate)
      : new Date().setHours(0, 0, 0, 0)
    const endDate = !isNil(options.endDate)
      ? Date.parse(options.endDate)
      : new Date().valueOf()
    googleFit.getDailyDistanceSamples(
      startDate,
      endDate,
      msg => {
        callback(msg, false)
      },
      res => {
        if (res.length > 0) {
          callback(false, prepareResponse(res, 'distance'))
        } else {
          callback('There is no any distance data for this period', false)
        }
      }
    )
  }

  getActivitySamples(options, callback) {
    googleFit.getActivitySamples(
      options.startDate,
      options.endDate,
      error => {
        callback(error, false)
      },
      res => {
        if (res.length > 0) {
          callback(false, res)
        } else {
          callback('There is no any distance data for this period', false)
        }
      }
    )
  }

  /**
   * Get the total calories per day over a specified date range.
   * @param {Object} options getDailyCalorieSamples accepts an options object containing:
   * required startDate: ISO8601Timestamp and endDate: ISO8601Timestamp
   * optional basalCalculation - {true || false} should we substract the basal metabolic rate averaged over a week
   * @param {Function} callback The function will be called with an array of elements.
   */

  getDailyCalorieSamples(options, callback) {
    const basalCalculation = options.basalCalculation !== false
    const startDate = Date.parse(options.startDate)
    const endDate = Date.parse(options.endDate)
    googleFit.getDailyCalorieSamples(
      startDate,
      endDate,
      basalCalculation,
      msg => {
        callback(msg, false)
      },
      res => {
        if (res.length > 0) {
          callback(false, prepareResponse(res, 'calorie'))
        } else {
          callback('There is no any calorie data for this period', false)
        }
      }
    )
  }

  getDailyNutritionSamples(options, callback) {
    const startDate = Date.parse(options.startDate)
    const endDate = Date.parse(options.endDate)
    googleFit.getDailyNutritionSamples(
      startDate,
      endDate,
      msg => {
        callback(msg, false)
      },
      res => {
        if (res.length > 0) {
          callback(false, prepareDailyResponse(res))
        } else {
          callback('There is no any nutrition data for this period', false)
        }
      }
    )
  }

  saveFood(options, callback) {
    options.date = Date.parse(options.date)
    googleFit.saveFood(
      options,
      msg => {
        callback(msg, false)
      },
      res => {
        callback(false, res)
      }
    )
  }

  /**
   * Query for weight samples. the options object is used to setup a query to retrieve relevant samples.
   * @param {Object} options  getDailyStepCountSamples accepts an options object containing unit: "pound"/"kg",
   *                          startDate: ISO8601Timestamp and endDate: ISO8601Timestamp.
   * @callback callback The function will be called with an array of elements.
   */

  getWeightSamples = (options, callback) => {
    const startDate = !isNil(options.startDate)
      ? Date.parse(options.startDate)
      : new Date().setHours(0, 0, 0, 0)
    const endDate = !isNil(options.endDate)
      ? Date.parse(options.endDate)
      : new Date().valueOf()
    googleFit.getWeightSamples(
      startDate,
      endDate,
      msg => {
        callback(msg, false)
      },
      res => {
        if (res.length > 0) {
          res = res.map(el => {
            if (el.value) {
              if (options.unit === 'pound') {
                el.value = KgToLbs(el.value) //convert back to pounds
              }
              el.startDate = new Date(el.startDate).toISOString()
              el.endDate = new Date(el.endDate).toISOString()
              return el
            }
          })
          callback(false, res.filter(day => !isNil(day)))
        } else {
          callback('There is no any weight data for this period', false)
        }
      }
    )
  }

  getHeightSamples(options, callback) {
    const startDate = Date.parse(options.startDate)
    const endDate = Date.parse(options.endDate)
    googleFit.getHeightSamples(
      startDate,
      endDate,
      msg => {
        callback(msg, false)
      },
      res => {
        if (res.length > 0) {
          callback(false, prepareResponse(res, 'value'))
        } else {
          callback('There is no any height data for this period', false)
        }
      }
    )
  }

  saveHeight(options, callback) {
    options.date = Date.parse(options.date)
    googleFit.saveHeight(
      options,
      msg => {
        callback(msg, false)
      },
      res => {
        callback(false, res)
      }
    )
  }

  saveWeight(options, callback) {
    if (options.unit == 'pound') {
      options.value = lbsAndOzToK({ pounds: options.value, ounces: 0 }) //convert pounds and ounces to kg
    }
    options.date = Date.parse(options.date)
    googleFit.saveWeight(
      options,
      msg => {
        callback(msg, false)
      },
      res => {
        callback(false, res)
      }
    )
  }

  deleteWeight = (options, callback) => {
    if (options.unit === 'pound') {
      options.value = lbsAndOzToK({ pounds: options.value, ounces: 0 }) //convert pounds and ounces to kg
    }
    googleFit.deleteWeight(
      prepareDeleteOptions(options),
      msg => {
        callback(msg, false)
      },
      res => {
        callback(false, res)
      }
    )
  }

  deleteHeight = (options, callback) => {
    googleFit.deleteWeight(
      prepareDeleteOptions(options),
      msg => {
        callback(msg, false)
      },
      res => {
        callback(false, res)
      }
    )
  }

  isAvailable(callback) {
    // true if GoogleFit installed
    googleFit.isAvailable(
      msg => {
        callback(msg, false)
      },
      res => {
        callback(false, res)
      }
    )
  }

  isEnabled(callback) {
    // true if permission granted
    googleFit.isEnabled(
      msg => {
        callback(msg, false)
      },
      res => {
        callback(false, res)
      }
    )
  }

  openFit() {
    googleFit.openFit()
  }

  observeSteps = callback => {
    const stepsObserver = DeviceEventEmitter.addListener(
      'StepChangedEvent',
      steps => callback(steps)
    )
    googleFit.observeSteps()
    this.eventListeners.push(stepsObserver)
  }

  observeHistory = callback => {
    const historyObserver = DeviceEventEmitter.addListener(
      'StepHistoryChangedEvent',
      steps => callback(steps)
    )
    this.eventListeners.push(historyObserver)
  }

  onAuthorize = callback => {
    const authObserver = DeviceEventEmitter.addListener(
      'GoogleFitAuthorizeSuccess',
      authorized => callback(authorized)
    )
    this.eventListeners.push(authObserver)
  }

  onAuthorizeFailure = callback => {
    const authFailedObserver = DeviceEventEmitter.addListener(
      'GoogleFitAuthorizeFailure',
      authorized => callback(authorized)
    )
    this.eventListeners.push(authFailedObserver)
  }

  unsubscribeListeners = () => {
    this.removeListeners()
  }

  getHeartRateSamples(options, callback) {
    const startDate = Date.parse(options.startDate)
    const endDate = Date.parse(options.endDate)
    googleFit.getHeartRateSamples(
      startDate,
      endDate,
      msg => {
        callback(msg, false)
      },
      res => {
        if (res.length > 0) {
          callback(false, prepareResponse(res, 'value'))
        } else {
          callback('There is no any heart rate data for this period', false)
        }
      }
    )
  }

  getBloodPressureSamples(options, callback) {
    const startDate = Date.parse(options.startDate)
    const endDate = Date.parse(options.endDate)
    googleFit.getBloodPressureSamples(
      startDate,
      endDate,
      msg => {
        callback(msg, false)
      },
      res => {
        if (res.length > 0) {
          callback(false, prepareResponse(res, 'value'))
        } else {
          callback('There is no any heart rate data for this period', false)
        }
      }
    )
  }

  /**
   * Get the sleep sessions over a specified date range.
   * @param {Object} options getSleepData accepts an options object containing required startDate: ISO8601Timestamp and endDate: ISO8601Timestamp.
   * @param {Function} callback The function will be called with an array of elements.
   */

  getSleepData = (options, callback) => {
    const startDate = !isNil(options.startDate)
      ? Date.parse(options.startDate)
      : new Date().setHours(0, 0, 0, 0)
    const endDate = !isNil(options.endDate)
      ? Date.parse(options.endDate)
      : new Date().valueOf()

    googleFit.getSleepData(
      startDate,
      endDate,
      msg => callback(true, msg),
      res => {
        if (res.length > 0) {
          callback(false, prepareResponse(res, 'value'))
        } else {
          callback('There is no sleep data for this period.', false)
        }
      }
    )
  }

}

export default new RNGoogleFit()

// Possible Scopes
export const Scopes = Object.freeze(PossibleScopes)

//Data types for food addition
export const MealType = Object.freeze({
  UNKNOWN: 0,
  BREAKFAST: 1,
  LUNCH: 2,
  DINNER: 3,
  SNACK: 4,
})

export const Nutrient = Object.freeze({
  /**
   * Calories in kcal
   * @type {string}
   */
  CALORIES: 'calories',
  /**
   * Total fat in grams.
   * @type {string}
   */
  TOTAL_FAT: 'fat.total',
  /**
   * Saturated fat in grams.
   * @type {string}
   */
  SATURATED_FAT: 'fat.saturated',
  /**
   * Unsaturated fat in grams.
   * @type {string}
   */
  UNSATURATED_FAT: 'fat.unsaturated',
  /**
   * Polyunsaturated fat in grams.
   * @type {string}
   */
  POLYUNSATURATED_FAT: 'fat.polyunsaturated',
  /**
   * Monounsaturated fat in grams.
   * @type {string}
   */
  MONOUNSATURATED_FAT: 'fat.monounsaturated',
  /**
   * Trans fat in grams.
   * @type {string}
   */
  TRANS_FAT: 'fat.trans',
  /**
   * Cholesterol in milligrams.
   * @type {string}
   */
  CHOLESTEROL: 'cholesterol',
  /**
   * Sodium in milligrams.
   * @type {string}
   */
  SODIUM: 'sodium',
  /**
   * Potassium in milligrams.
   * @type {string}
   */
  POTASSIUM: 'potassium',
  /**
   * Total carbohydrates in grams.
   * @type {string}
   */
  TOTAL_CARBS: 'carbs.total',
  /**
   * Dietary fiber in grams
   * @type {string}
   */
  DIETARY_FIBER: 'dietary_fiber',
  /**
   * Sugar amount in grams.
   * @type {string}
   */
  SUGAR: 'sugar',
  /**
   * Protein amount in grams.
   * @type {string}
   */
  PROTEIN: 'protein',
  /**
   * Vitamin A amount in International Units (IU).
   * @type {string}
   */
  VITAMIN_A: 'vitamin_a',
  /**
   * Vitamin C amount in milligrams.
   * @type {string}
   */
  VITAMIN_C: 'vitamin_c',
  /**
   * Calcium amount in milligrams.
   * @type {string}
   */
  CALCIUM: 'calcium',
  /**
   * Iron amount in milligrams
   * @type {string}
   */
  IRON: 'iron',
})

/*
TODO: Add food example to readme

same as here: https://developers.google.com/fit/scenarios/add-nutrition-data

import GoogleFit, {Nutrient, MealType, FoodIntake, WeightSample} from "react-native-google-fit";
...
    addFoodExample(): Promise<void> {
        return new Promise<void>((resolve, reject): void => {
            const options = {
                mealType: MealType.BREAKFAST,
                foodName: "banana",
                date: moment().format(), //equals to new Date().toISOString()
                nutrients: {
                    [Nutrient.TOTAL_FAT]: 0.4,
                    [Nutrient.SODIUM]: 1,
                    [Nutrient.SATURATED_FAT]: 0.1,
                    [Nutrient.PROTEIN]: 1.3,
                    [Nutrient.TOTAL_CARBS]: 27.0,
                    [Nutrient.CHOLESTEROL]: 0,
                    [Nutrient.CALORIES]: 105,
                    [Nutrient.SUGAR]: 14,
                    [Nutrient.DIETARY_FIBER]: 3.1,
                    [Nutrient.POTASSIUM]: 422,
                }
            } as FoodIntake;

            GoogleFit.saveFood(options, (err: boolean) => {
                if (!err) {
                    resolve();
                } else {
                    reject();
                }
            });
        });
    }
*/
